Normalize the scheme and host in the URL parser

jeremycline opened the pull request:
I don't know that this is the most rigorous way to fix the problem of case-sensitivity, but it seemed like a nice centralized place to normalize things.

According to RFC 3986, the scheme and the host portion of the authority
in a URI are case-insensitive. ``urllib3`` uses the scheme in particular
as a key for various dictionaries, and oddly-cased schemes can break
this. This patch modifies the Url class to normalize both the scheme and
host.

fixes #833

haikuginger commented:
I'm 👍; RFC-7230 affirms that scheme and host are case-insensitive.

Lukasa commented:
This looks reasonable enough to me, but @sigmavirus24 is our resident URL expert, so I'd love it if he weighed in here before I merged this.

sigmavirus24 commented:
@haikuginger the correct RFC to reference is 3986. That said, I'd feel more comfortable if we had tests affirming that the userinfo isn't changed ever. That part *is* case-sensitive. (I think urllib3 calls this part 'auth' as in authentication, not authority as you'll find the (userinfo@host:port) portion called in 3986.)

haikuginger commented:
> @haikuginger the correct RFC to reference is 3986.

Thanks! I think I might have confused the relationship of RFC-7230 to 3986 with the relationship of 3986 to 2396.

I agree that a test to ensure that other URL components aren't ever modified is a good idea.

jeremycline commented:
I shall make it so. Thanks for the feedback!

jeremycline commented:
@sigmavirus24 Does 90fa353 cover everything?

sigmavirus24 commented:
Thanks @jeremycline! Looks good to me. :+1:

Lukasa commented:
Hurrah, thanks both! :sparkles: :cake: :sparkles: @jeremycline Do you mind updating CHANGES with this? Frustratingly I have to take an extra step to do it myself because I don't have admin on this repo, so it's easier if your PR does it for me. =D

jeremycline commented:
@Lukasa No problem. I've rebased, squashed the two commits for tests into one, and added a changelog entry. Let me know if the changelog isn't what you had in mind and I'll be happy to tweak it as necessary.

Lukasa commented:
Hurrah, thanks @jeremycline! :sparkles:
